pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -308,28 +371,79 @@ func (s *Scheduler) add(ctx context.Context, pod *corev1.Pod) error {
+		}
+
 		instanceTypes := nodeClaimTemplate.InstanceTypeOptions
 		// if limits have been applied to the nodepool, ensure we filter instance types to avoid violating those limits
 		if remaining, ok := s.remainingResources[nodeClaimTemplate.NodePoolName]; ok {
 			instanceTypes = filterByRemainingResources(instanceTypes, remaining)
 			if len(instanceTypes) == 0 {
-				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed limits for nodepool: %q", nodeClaimTemplate.NodePoolName))
+				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed limits for nodepool %q", nodeClaimTemplate.NodePoolName))
 				continue
 			} else if len(nodeClaimTemplate.InstanceTypeOptions) != len(instanceTypes) {
-				log.FromContext(ctx).V(1).WithValues("NodePool", klog.KRef("", nodeClaimTemplate.NodePoolName)).Info(fmt.Sprintf("%d out of %d instance types were excluded because they would breach limits",
-					len(nodeClaimTemplate.InstanceTypeOptions)-len(instanceTypes), len(nodeClaimTemplate.InstanceTypeOptions)))
+				log.FromContext(ctx).V(1).WithValues(
+					"NodePool", klog.KRef("", nodeClaimTemplate.NodePoolName),
+				).Info(fmt.Sprintf(
+					"%d out of %d instance types were excluded because they would breach limits",
+					len(nodeClaimTemplate.InstanceTypeOptions)-len(instanceTypes),
+					len(nodeClaimTemplate.InstanceTypeOptions),
+				))
 			}
 		}
-		nodeClaim := NewNodeClaim(nodeClaimTemplate, s.topology, s.daemonOverhead[nodeClaimTemplate], instanceTypes)
+
+		nodeClaim := NewNodeClaim(nodeClaimTemplate, s.topology, s.daemonOverhead[nodeClaimTemplate], instanceTypes, s.reservationManager, s.reservedOfferingMode)
 		if err := nodeClaim.Add(pod, s.cachedPodData[pod.UID]); err != nil {
-			nodeClaim.Destroy() // Ensure we cleanup any changes that we made while mocking out a NodeClaim
-			errs = multierr.Append(errs, fmt.Errorf("incompatible with nodepool %q, daemonset overhead=%s, %w",
+			nodeClaim.Destroy()
+			if IsReservedOfferingError(err) {
+				errs = multierr.Append(errs, fmt.Errorf(
+					"compatible with nodepool %q but failed to add pod while adhering to reservation fallback policy, %w",
+					nodeClaimTemplate.NodePoolName,
+					err,
+				))
+				compatibleReservedNCT = nodeClaimTemplate
+			} else {
+				errs = multierr.Append(errs, fmt.Errorf(
+					"incompatible with nodepool %q, daemonset overhead=%s, %w",
+					nodeClaimTemplate.NodePoolName,
+					resources.String(s.daemonOverhead[nodeClaimTemplate]),
+					err,
+				))
+			}
+			continue
+		}
+
+		// TODO: If necessary, could mark the reserved offering status in the scheduling.NodeClaim rather than crawling the
+		// offerings. Granted, this won't be an issue unless there is a compatible reserved NodeClaimTemplate since we short
+		// circuit the conditional.
+		if s.reservedOfferingMode == ReservedOfferingModeStrict && compatibleReservedNCT != nil && !lo.ContainsBy(nodeClaim.InstanceTypeOptions, func(it *cloudprovider.InstanceType) bool {
```

Comment:
Maybe I'm crazy but this seems pretty complicated for me to reason about -- does this problem just evaporate if we don't worry about equal-weight NodePools?

Reply:
Yeah, pretty much. I still think this might be worth keeping in, given the latency advantages, but I can be convinced otherwise. If we do decide to keep it in though, I can work on some improved comments and variable names to hopefully make things more clear.

Reply:
Discussed offline, this is something we can consider bringing back if we see issues with scheduling latency, but given the limited benefit and the increased complexity we're going to drop the changes for now.